From 0ba7fe5a10874bc8915b40dc50154b3c44fa0cf5 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Sun, 22 Jun 2025 18:00:45 +0200
Subject: [PATCH 0983/1010] evl/net: skb: fix in-band release path

free_oob_skb() may receive in-band buffer shells (!skb_is_oob())
conveying oob storage, those were wrongly released to the oob pool so
far, causing OOM eventually.

Relying on the latest Dovetail addition to __kfree_skb() for handling
re-entry from the in-band stage of skbs already released by the oob
core, relay those buffers to finalize_skb_inband instead, so that
their head state is properly flushed and the buffer shell is released
to the proper kmem cache.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 kernel/evl/net/skb.c | 38 ++++++++++++++++++++++++++++++--------
 1 file changed, 30 insertions(+), 8 deletions(-)

diff --git a/kernel/evl/net/skb.c b/kernel/evl/net/skb.c
index 8ec69e644c..62990c456f 100644
--- a/kernel/evl/net/skb.c
+++ b/kernel/evl/net/skb.c
@@ -215,12 +215,6 @@ static void __free_evl_skb(struct sk_buff *skb, struct net_device *dev)
 	struct evl_netdev_state *est = dev->oob_state.estate;
 	unsigned long flags;
 
-	/* If the data storage is still shared, don't release it. */
-	if (skb->cloned &&
-	    atomic_sub_return(skb->nohdr ? (1 << SKB_DATAREF_SHIFT) + 1 : 1,
-			      &shinfo->dataref))
-		goto put_skb;
-
 	/*
 	 * Attempt to release the heading data to the originating page
 	 * pool if any.
@@ -228,8 +222,24 @@ static void __free_evl_skb(struct sk_buff *skb, struct net_device *dev)
 	if (unlikely(!skb->head))
 		goto put_skb;
 
+	/* If the data storage is still shared, don't release it. */
+	if (skb->cloned &&
+	    atomic_sub_return(skb->nohdr ? (1 << SKB_DATAREF_SHIFT) + 1 : 1,
+			      &shinfo->dataref))
+		goto put_skb;
+
+	EVL_WARN_ON(NET, atomic_read(&shinfo->dataref) < 0);
+
 	skb_pp_recycle(skb, skb->head);
 
+	/*
+	 * Make sure skb_release_all() won't attempt to release the
+	 * data portion - we just did so - if we need to relay the
+	 * buffer downstream for finalization (see finalize_skb_inband
+	 * -> __kfree_skb -> skb_release_all).
+	 */
+	skb->head = NULL;
+
 	/*
 	 * Wake up any thread waiting for buffer space to send to the
 	 * device we are releasing the page to.
@@ -244,8 +254,20 @@ static void __free_evl_skb(struct sk_buff *skb, struct net_device *dev)
 	evl_signal_poll_events(&est->poll_head,	POLLOUT|POLLWRNORM);
 
 put_skb:
-	EVL_WARN_ON(NET, atomic_read(&shinfo->dataref) < 0);
-	put_oob_skb(skb);
+	/*
+	 * We may receive skbs obtained from the oob pool or some
+	 * in-band kmem cache indifferently, depending on where the
+	 * skb originates from (e.g. RX path of an oob-capable device
+	 * or not). If the skb shell belongs to the oob pool, release
+	 * it immediately.  Otherwise we need a little help from the
+	 * in-band stack for finalizing this buffer, relay it
+	 * downstream for release so that the head state is flushed as
+	 * well.
+	 */
+	if (skb_is_oob(skb))
+		put_oob_skb(skb);
+	else
+		free_inband_skb(skb);
 }
 
 /*
-- 
2.30.2

